client/kv: add sequence numbers to individual Requests

nvb opened the pull request:
Fixes #25007.

This change adds a sequence number to each individual Request in a
BatchRequest, addressing a number of problems laid out in the referenced
issue and paving the way for future changes including:
- pipelined writes
- parallel commits
- kv-level idempotency

### Approach

This change takes a simple approach to augmenting this new state - it adds
a sequence number to all Request implementations by adding it as a field
to `RequestHeader`. This is straightforward and easy to reason about but
provides a few downsides. The most serious is that Requests now carry
transactional state, which means that instances of Requests are more
difficult to send multiple times. The biggest place where this shows up
is in testing. It also forces tests that handle transactions themselves
to be a little more explicit about sequence numbers, but this might not
be a bad thing.

Interestingly, a side effect of this change is that it fixes a previous
race where concurrent requests sent through `client.Txn` could end up
with the same sequence number because these seq nums used to be set in
DistSender. This race is harmless today only because of our conservative
use of transaction concurrency. Because this change fixes the race, it
has the effect of making concurrent use of client.Txn safer.

### Migration

The migration for this change is fairly straightforward. 2.1 nodes will
now begin supplying Request-scoped sequence numbers on all transactional
BatchRequests. Clients will also continue to send BatchRequest-scoped
sequence numbers, which will always be the same as the maximum seq num of
all individual Requests in the pre-split (by DistSender) BatchRequest. This
is important because it will ensure backward compatibility with 2.0 nodes.
One small effect of this is that BatchRequests that are split between
ranges will now appear have the same sequence number to 2.0 nodes. This
is ok, because these BatchRequests necessarily have disjoint key sets.

When a server observes a BatchRequest-scoped sequence number (which is
identified by a lack of Request-scoped sequence numbers) it assigns a
BatchIndex, as it used to. Servers continue to handle BatchIndexes in
existing intents as usual.

Each Request having its own Sequence provides the same safety guarantees
as each BatchRequest having a Sequence and each Request being distinguished
using the BatchIndex field. Furthermore, a transaction can theoretically
switch back and forth between the two approaches without issue.

### Performance Considerations

The change has two small performance implications, both of which I believe will
be negligible and which offset each other (this assumption will be confirmed
before merging, if the proposal gets that far). The first implication is that
every individual Request will now include a varint sequence number. These numbers
will almost always be very low, so they'll usually only require a byte or two extra
when a BatchRequest is on the wire. Still, this means that the network overhead of
KV requests will increase slightly. The other implication is that we'll no longer
persist a BatchIndex in intents since the change collapses the batch index
dimension into the sequence number dimension. This should save a byte or two on
any intent that was not created by the first Request of a Batch (because
BatchIndex was 0-indexed). This should reduce network usage and disk usage
beneath Raft).

A future optimization could be to exclude the sequence number on the last
Request in a BatchRequest, since it will always be the same as the sequence
number in the Header.Txn. It's not clear whether this optimization is worth it
though, and that depends on the average number of Requests in a BatchRequest.
Additionally, it would be best to hold off until 2.2 to make the optimization,
since making it now would complicate the upgrade story. This is because any
Request without a sequence number is currently used as an indication of
legacy-style sequencing.

An alternate optimization could be to strip the sequence number from
Header.Txn before sending it on the wire once backward compatibility
is no longer a concern.

Release note: None

nvb committed:
storage: don't increase sequence number on WriteIntentError

This change removes a call to `SetNewRequest` in `Store.Send's` retry
loop. The call was being used to increase the sequence counter on a
batch to "avoid getting caught in replay protection on retry", but this
didn't make sense because the presence of a `WriteIntentError` implies
that no intents were written. This may have been necessary previously,
before we removed the ability to write through Raft when an error was
experienced.

This change ensures that all sequence number decisions are now made by
the transaction coordinator client.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25728)
<!-- Reviewable:end -->

bdarnell commented:
Reviewed 12 of 15 files at r2, 27 of 27 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[pkg/internal/client/txn.go, line 955 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25728#-LD2NH0v1ops19rX0W1a:-LD2NH0v1ops19rX0W1b:b-p0xqdb) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfdf51ffb59b4a55bd5801689d5cfb2da1d375f/pkg/internal/client/txn.go#L955)):*
> ```Go
> 		// transaction would need to be strengthened elsewhere (e.g. by the
> 		// transport layer).
> 		if txn.db.st.Version.IsActive(cluster.VersionIndividualSequenceNumbers) {
> ```

I'd expect 2.1 to always send a sequence number, and the version check would be on the receiving side. Why did you do it this way?

Also, the setting could change in the middle of a single transaction, which seems surprising. 

---

*[pkg/roachpb/api.proto, line 55 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25728#-LD2LE8KESvO35_DUW10:-LD2LE8KESvO35_DUW11:be2x0nz) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfdf51ffb59b4a55bd5801689d5cfb2da1d375f/pkg/roachpb/api.proto#L55)):*
> ```protobuf
> 
> // SeqNum is provided with every storage node request.
> message SeqNum {
> ```

While we're adding a new field, we should transition to a proper RequestHeader message so that the next time we have to add a field we have a place to put it. We could do this in two ways:

- Add a message containing (`start_key`, `end_key`, `seq_num`) and immediately swap all references to `Span header` to this new message. Add an accessor method to construct a Span from this type.
- Add a message containing (`span`, `seq_num`)  and add it alongside the `Span header` in all request objects. Support both formats in 2.1 with the intention of removing the old span header from 2.2.

---

*[pkg/roachpb/api.proto, line 1093 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25728#-LD2L4kmBol4IIgTZijf:-LD2L4kn-bD2Oc7lEUQl:baju99w) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d0c09aee9ff49ff0e6ba09bac86f5da90f9b63c/pkg/roachpb/api.proto#L1093)):*
> ```protobuf
>     bytes sha512 = 4;
>   }
>   Span header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
> ```

Add `reserved 1` whenever you remove a field like this.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25728)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/internal/client/txn.go, line 955 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25728#-LD2NH0v1ops19rX0W1a:-LD3DGqo-DfHRA-QyRdv:bgvznb9) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfdf51ffb59b4a55bd5801689d5cfb2da1d375f/pkg/internal/client/txn.go#L955)):*
> Why did you do it this way?

I did it this way so that we could stop adjusting sequence numbers in `DistSender` at the same time that we started setting sequence numbers here. This is important to avoid races where sequence numbers are set in two places under two differtent locks.

> Also, the setting could change in the middle of a single transaction, which seems surprising.

Yeah, I touched upon this in the commit message's "Migration" section (see "One interesting case..."). It could be surprising, but it should be completely safe. I don't think it's worth adding extra state to `client.Txn` to avoid it.

---

*[pkg/roachpb/api.proto, line 55 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25728#-LD2LE8KESvO35_DUW10:-LD3Dra73PnJj2E19Inv:b-nhebt8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfdf51ffb59b4a55bd5801689d5cfb2da1d375f/pkg/roachpb/api.proto#L55)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

While we're adding a new field, we should transition to a proper RequestHeader message so that the next time we have to add a field we have a place to put it. We could do this in two ways:

- Add a message containing (`start_key`, `end_key`, `seq_num`) and immediately swap all references to `Span header` to this new message. Add an accessor method to construct a Span from this type.
- Add a message containing (`span`, `seq_num`)  and add it alongside the `Span header` in all request objects. Support both formats in 2.1 with the intention of removing the old span header from 2.2.
</blockquote></details>

I touched on this in the commit message's "Approach" section (see "If there was a RequestHeader message type..."). I began by looking into the second approach but figured it would have an even larger migration than this PR's current approach and would be somewhat risky. 

I wasn't aware that the first approach was even possible, but it's the approach I would use if it's safe. To confirm, this is safe because two proto messages are interchangeable if they have the same structure (types & field numbers of shared fields), even if they have different names, right? And adding the third field is just a normal proto migration as long as it has a previously unused field number?

---

*[pkg/roachpb/api.proto, line 1093 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25728#-LD2L4kmBol4IIgTZijf:-LD3Dlbx0cTRRkOTFtFW:b-2f2gbl) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d0c09aee9ff49ff0e6ba09bac86f5da90f9b63c/pkg/roachpb/api.proto#L1093)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Add `reserved 1` whenever you remove a field like this.
</blockquote></details>

I didn't remove this field, just moved it up.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25728)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/internal/client/txn.go, line 955 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25728#-LD2NH0v1ops19rX0W1a:-LD8EK3Q4dbTgAub67AO:b-2zomsq) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfdf51ffb59b4a55bd5801689d5cfb2da1d375f/pkg/internal/client/txn.go#L955)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> Why did you do it this way?

I did it this way so that we could stop adjusting sequence numbers in `DistSender` at the same time that we started setting sequence numbers here. This is important to avoid races where sequence numbers are set in two places under two differtent locks.

> Also, the setting could change in the middle of a single transaction, which seems surprising.

Yeah, I touched upon this in the commit message's "Migration" section (see "One interesting case..."). It could be surprising, but it should be completely safe. I don't think it's worth adding extra state to `client.Txn` to avoid it.
</blockquote></details>

OK, I should've gone back to re-read the commit message. This seems OK, but will it complicate the use of the yet-to-be-developed features that rely on per-request sequence numbers?  (I'm not sure yet how those will fit together)

---

*[pkg/roachpb/api.proto, line 55 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25728#-LD2LE8KESvO35_DUW10:-LD8AohJ5B4BjDt9ukhd:bemnhi) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfdf51ffb59b4a55bd5801689d5cfb2da1d375f/pkg/roachpb/api.proto#L55)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I touched on this in the commit message's "Approach" section (see "If there was a RequestHeader message type..."). I began by looking into the second approach but figured it would have an even larger migration than this PR's current approach and would be somewhat risky. 

I wasn't aware that the first approach was even possible, but it's the approach I would use if it's safe. To confirm, this is safe because two proto messages are interchangeable if they have the same structure (types & field numbers of shared fields), even if they have different names, right? And adding the third field is just a normal proto migration as long as it has a previously unused field number?
</blockquote></details>

Yes, proto messages are interchangeable regardless of name if the have compatible types and tag numbers (and the JSON encoding is not used). 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25728)*
<!-- Sent from Reviewable.io -->

nvb committed:
client/kv: add sequence numbers to individual Requests

Fixes #25007.

This change adds a sequence number to each individual Request in a
BatchRequest, addressing a number of problems laid out in the referenced
issue and paving the way for future changes including:
- pipelined writes
- parallel commits
- kv-level idempotency

_### Approach

This change takes a simple approach to augmenting this new state - it adds
a sequence number to all Request implementations by adding it as a field
to `RequestHeader`. This is straightforward and easy to reason about but
provides a few downsides. The most serious is that Requests now carry
transactional state, which means that instances of Requests are more
difficult to send multiple times. The biggest place where this shows up
is in testing. It also forces tests that handle transactions themselves
to be a little more explicit about sequence numbers, but this might not
be a bad thing.

Interestingly, a side effect of this change is that it fixes a previous
race where concurrent requests sent through `client.Txn` could end up
with the same sequence number because these seq nums used to be set in
DistSender. This race is harmless today only because of our conservative
use of transaction concurrency. Because this change fixes the race, it
has the effect of making concurrent use of client.Txn safer.

_### Migration

The migration for this change is fairly straightforward. 2.1 nodes will
now begin supplying Request-scoped sequence numbers on all transactional
BatchRequests. Clients will also continue to send BatchRequest-scoped
sequence numbers, which will always be the same as the maximum seq num of
all individual Requests in the pre-split (by DistSender) BatchRequest. This
is important because it will ensure backward compatibility with 2.0 nodes.
One small effect of this is that BatchRequests that are split between
ranges will now appear have the same sequence number to 2.0 nodes. This
is ok, because these BatchRequests necessarily have disjoint key sets.

When a server observes a BatchRequest-scoped sequence number (which is
identified by a lack of Request-scoped sequence numbers) it assigns a
BatchIndex, as it used to. Servers continue to handle BatchIndexes in
existing intents as usual.

Each Request having its own Sequence provides the same safety guarantees
as each BatchRequest having a Sequence and each Request being distinguished
using the BatchIndex field. Furthermore, a transaction can theoretically
switch back and forth between the two approaches without issue.

_### Performance Considerations

The change has two small performance implications, both of which I believe will
be negligible and which offset each other (this assumption will be confirmed
before merging, if the proposal gets that far). The first implication is that
every individual Request will now include a varint sequence number. These numbers
will almost always be very low, so they'll usually only require a byte or two extra
when a BatchRequest is on the wire. Still, this means that the network overhead of
KV requests will increase slightly. The other implication is that we'll no longer
persist a BatchIndex in intents since the change collapses the batch index
dimension into the sequence number dimension. This should save a byte or two on
any intent that was not created by the first Request of a Batch (because
BatchIndex was 0-indexed). This should reduce network usage and disk usage
beneath Raft).

A future optimization could be to exclude the sequence number on the last
Request in a BatchRequest, since it will always be the same as the sequence
number in the Header.Txn. It's not clear whether this optimization is worth it
though, and that depends on the average number of Requests in a BatchRequest.
Additionally, it would be best to hold off until 2.2 to make the optimization,
since making it now would complicate the upgrade story. This is because any
Request without a sequence number is currently used as an indication of
legacy-style sequencing.

An alternate optimization could be to strip the sequence number from
Header.Txn before sending it on the wire once backward compatibility
is no longer a concern.

Release note: None

nvb committed:
roachpb: create RequestHeader message type, split from Span

This change creates a RequestHeader message type and uses this new type
to replace Span as the Header field in all Request message types. The
message is given the same field types and numbers as Span, so it is
wire compatible without a migration.

Future commits will add Request-specific fields to this message type,
which is why the message type is diverging from Span.

Release note: None

nvanbenschoten commented:
I just addressed both comments but still need to go back and update the commit message. I'll also do a bit of benchmarking (using tpccbench!) to ensure that the extra sequence numbers have a negligible perf cost.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/internal/client/txn.go, line 955 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25728#-LD2NH0v1ops19rX0W1a:-LD9s_NUACnp9T75U6wS:b-f5b53l) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfdf51ffb59b4a55bd5801689d5cfb2da1d375f/pkg/internal/client/txn.go#L955)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

OK, I should've gone back to re-read the commit message. This seems OK, but will it complicate the use of the yet-to-be-developed features that rely on per-request sequence numbers?  (I'm not sure yet how those will fit together)
</blockquote></details>

Yeah, I think you're right. It should be fine to set the BatchRequest-scoped sequence number to the max of all individual sequence numbers in `client.Txn` and no longer set it in DistSender at all. If a BatchRequest is split into two batches between two ranges (which necessarily have disjoint key sets) then it is fine for them to have the same sequence number.

This allows 2.1 clients beneath `client.Txn` to always assume that Requests have individual sequence numbers, which will make their development more straightforward

---

*[pkg/roachpb/api.proto, line 55 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25728#-LD2LE8KESvO35_DUW10:-LD9tBnH8Tq9Vqk5Xpea:b-w0vzz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfdf51ffb59b4a55bd5801689d5cfb2da1d375f/pkg/roachpb/api.proto#L55)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yes, proto messages are interchangeable regardless of name if the have compatible types and tag numbers (and the JSON encoding is not used). 
</blockquote></details>

Done. That turned out much nicer, other than the huge new refactor commit.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25728#-:-LD9sQ1FCFC__h3DzdaL:b-3128hd)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 81 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25728#-:-LDCmMM82RWbX_WIgUYF:bnfp4nl)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
TFTR @bdarnell. I'm not going to merge the PR this week because of its stability implications. Also, as promised I'll verify that this doesn't have any performance effects by running it against tpccbench.

@tschottdorf no need for a full code review, but I'm curious if the approach here matches your expectations (see commit messages).

tschottdorf commented:
It does, thanks!

On Wed, May 23, 2018, 10:19 PM Nathan VanBenschoten <
notifications@github.com> wrote:

> TFTR @bdarnell <https://github.com/bdarnell>. I'm not going to merge the
> PR this week because of its stability implications. Also, as promised I'll
> verify that this doesn't have any performance effects by running it against
> tpccbench.
>
> @tschottdorf <https://github.com/tschottdorf> no need for a full code
> review, but I'm curious if the approach here matches your expectations (see
> commit messages).
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/25728#issuecomment-391482744>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135At0eoo15effQtiAfr12emgVSrjNks5t1cQ-gaJpZM4UF75U>
> .
>
-- 

-- Tobias

nvanbenschoten commented:
bors r+
